### ISSUE TITLE
Definition of hyperparameter equality

### ISSUE THREAD

**Issue by mfeurer** [automl/ConfigSpace#214]
Currently, the definition of hyperparameter equality is not equal for all hyperparameter types. For some we take the default value into account when comparing hyperparameters for equality, for others we don't. We need to unify. Are there any pros/cons for either decision?

From the top of my head I don't see a reason to include the default value into an equality comparison except for the fact that SMAC makes use of it.

**Comment by mlindauer** [comment]
Sorry, I don't understand this. Could you please give an example?

**Comment by eddiebergman** [comment]
I would personally expect the default value not to matter so equality should be purely based on its range of outputs.

**Comment by mfeurer** [comment]
> Sorry, I don't understand this. Could you please give an example?

Sure, for a categorical hyperparameter the definition of equality is
```
return (
    self.name == other.name and
    set(self.choices) == set(other.choices)
)
```
and for a uniform float it is
```
(
    self.name == other.name and
    self.default_value == other.default_value and
    self.lower == other.lower and
    self.upper == other.upper and
    self.log == other.log and
    self.q == other.q
)
```
which, most importantly, contains the default.

**Comment by mlindauer** [comment]
So, the underlying question is whether two configuration spaces are the same? 
I would argue that the default should be part of this equality check because providing information about different defaults could lead to different optimization trajectories (under the assumption that an optimizer makes use of it).

**Comment by KEggensperger** [comment]
I'd say the default value is not relevant for comparing two hyperparameters. I'd even say the name is not relevant ...

However, I can see use cases where this could matter and would raise a warning if the names or default values are not the same.

**Comment by eddiebergman** [comment]
I go back on my opinion after reading your comment @mlindauer , I think the main users of ConfigSpace are people interested in running Optimizers.

I would imagine this equality check would mainly happen when someone wants to compare two `ConfigurationSpace` objects which would then go and recursively check each hyperparameter. Given two Config spaces say they are "equal" and a deterministic optimizer, you would expect the same result, which is not the case given a different default value.

This is probably relatively minor in the grand scheme of things but I agree with Marius

**Comment by dengdifan** [comment]
if we consider that from an optimization perspective, then maybe random seed also needs to be considered. In practice, we might expect that two equal configspaces are derived from the same source (either by copying from one raw cs or reading from a file on the disk). Then default value should be a good indicator showing the equality of two configspace.

**Comment by mfeurer** [comment]
Fixed via #232
